core/trino-main/src/main/java/io/trino/execution/QueryExecution.java

```diff
@@ -99,22 +100,27 @@ interface QueryExecutionFactory<T extends QueryExecution>
      */
     class QueryOutputInfo
     {
-        private final List<String> columnNames;
+        private final List<ColumnInfo> columnInfos;
         private final List<Type> columnTypes;
         private final Queue<ExchangeInput> inputsQueue;
         private final boolean noMoreInputs;
 
-        public QueryOutputInfo(List<String> columnNames, List<Type> columnTypes, Queue<ExchangeInput> inputsQueue, boolean noMoreInputs)
+        public QueryOutputInfo(List<ColumnInfo> columnInfos, List<Type> columnTypes, Queue<ExchangeInput> inputsQueue, boolean noMoreInputs)
         {
-            this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
+            this.columnInfos = ImmutableList.copyOf(requireNonNull(columnInfos, "columnInfos is null"));
             this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
             this.inputsQueue = requireNonNull(inputsQueue, "inputsQueue is null");
             this.noMoreInputs = noMoreInputs;
         }
 
+        public List<ColumnInfo> getColumnInfos()
+        {
+            return columnInfos;
+        }
+
         public List<String> getColumnNames()
         {
-            return columnNames;
+            return columnInfos.stream().map(ColumnInfo::name).collect(toImmutableList());
```

Comment:
This is executed every time `QueryInfo` is constructed.
This might be not a big deal, but i am not sure.
Can you move column name list extraction to `QueryOutputInfo` constructor?